backend/app/analysis/proximity.py

```diff
@@ -0,0 +1,85 @@
+import nltk
+from collections import Counter
+from more_itertools import windowed
+
+from ..models import (
+    Document,
+    Gender,
+)
+
+
+def run_analysis(doc_ids, gender_ids, word_window):
+    """
+    Generates a dictionary of dictionaries for each `Document` object. Each dictionary maps a `Gender` to a word count
+    of words within a specified window of that `Gender`'s pronouns.
+
+    :param doc_ids: A list of ints representing `Document` ids
+    :param gender_ids: A list of ints representing `Gender` ids
+    :param word_window: An integer describing the number of words to look at of each side of a gendered word
+
+    :return: A dict mapping `Document` ids to a dict mapping strings (`Gender` labels) to a `Counter` instance.
+    """
+    results = {}
+
+    for key in doc_ids:
+        results[key] = generate_gender_token_counters(
+            Document.objects.values_list('tokenized_text', flat=True).filter(pk=key),
+            gender_ids,
+            word_window
+        )
+
+    return results
+
+
+def generate_gender_token_counters(text_query, gender_ids, word_window):
+    """
+    Generates a dictionary mapping `Gender`s to a word count of words within a specified window of the `Gender`'s
+    pronouns.
+
+    :param text_query: An unevaluated, length-1 `QuerySet` that returns a list of strings when evaluated
+    :param gender_ids: A list of ints representing `Gender` ids
+    :param word_window: An integer describing the number of words to look at of each side of a gendered word
+
+    :return: A dict mapping strings (`Gender` labels) to a `Counter` instance.
+
+    """
+
+    results = {}
+
+    for gender_id in gender_ids:
+        doc_result = generate_token_counter(text_query, gender_id, word_window)
+        results.update({Gender.objects.values_list('label', flat=True).get(pk=gender_id): doc_result})
```

Comment:
Good use of lazy evaluation! I was wondering if it may be more efficient to add in dictionary entries using `dict[key] = value` (see below) because your results dictionary is empty and you are adding new entries as you iterate over the IDs. From my understanding, `dict.update()` looks to see if the key already exists in the dictionary and if it does it overwrites that key-value pair or else it just adds the dictionary entry. Let me know what you think!
```suggestion
       gender_label = Gender.objects.values_list('label', flat=True).get(pk=gender_id)
       results[gender_label] = doc_result
```

Reply:
Thanks for the suggestion! I changed that line as you suggested, but I also just hit the database to load the `Gender` object in `generate_gender_token_counters` instead of `generate_token_counters`. After talking with @MBJean, we concluded that each `Gender` model is slim enough to hit the database earlier rather than later and save ourselves an extra query.